File: ConsoleVideoPlayer.MediaProcessor/PixelLookup.fs
module ConsoleVideoPlayer.MediaProcessor.PixelLookup

open SkiaSharp

type LookupData =
    {Image: SKBitmap
     // EW
     mutable PixelCache: SKColor array option}

let makeLookupData image = {Image = image; PixelCache = None}

let colourAtCoord lookup x y =
    let pixelCache =
        match lookup.PixelCache with
        | Some p -> p
        | None -> lookup.Image.Pixels

    // this is not nice but cache gotta cache
    lookup.PixelCache <- Some pixelCache

    pixelCache.[lookup.Image.Width * y + x]

File: ConsoleVideoPlayer.Img2Text/PixelLookup.cs
using SkiaSharp;

namespace ConsoleVideoPlayer.Img2Text;

public class PixelLookup
{
	public PixelLookup(string path) => _image = SKBitmap.Decode(path);

	private readonly SKBitmap _image;

	public (int width, int height) Dimensions => (_image.Width, _image.Height);

	private SKColor[]? _pixelsCache;
	private SKColor[]  Pixels => _pixelsCache ??= _image.Pixels;

	public SKColor ColourAtCoord(int x, int y) => Pixels[_image.Width * y + x];
}